### DISCUSSION TITLE
Build-time only attributes

### DISCUSSION THREAD

**Discussion by mgravell** [dotnet/csharplang#4715] [Language Ideas] [answered]
Attributes. We use them for lots of things, but some attributes aren't used at runtime - they are only used during build, either by the compiler, or (more recently) by generators. I envisage the use of generators increasing, and with it: the use of attributes to help influence intent.

But: lots of these attributes would have no utility at runtime. So: why emit them?

# Justification

- smaller IL / assembly size
- entire types can be dropped by the linker
- may often allow assembly references and possibly package references to be dropped during build
- fewer symbols (like literal strings from attributes) in the metadata tables
- lower impact from tools than scan all members looking at attributes

# Applicability

- compiler attributes like SkipLocalsInitAtrribute ModuleInitializerAttribute, EnumeratorCancellationAttribute, etc; not used except by the compiler itself
- external attributes used by generators; for example, Dapper AOT intends to use attributes to express intent re SQL strings (and a whole bunch of other things); those attributes can be burned - they aren't used

# Proposed design

- add a new attribute in System.Runtime.CompilerServices, for example BuildOnlyAttribute or BuildOnlyAttributeAttribute (naming is hard!)
- all attributes would continue to exist in the semantic model for inspection by the compiler and generators
- when emitting final IL, attributes marked with this type simply aren't written

# Questions

- should DEBUG builds always emit them?

# Example

Consider a `partial` method implemented by a generator, with attributes that the generator uses to understand the intent:

``` c#
[Bar("some long string"), Blap]
public partial int Foo(...);
```

Now consider that BarAttribute is marked build-time-only. The final IL would be for:

``` c#
[Blap]
// Plus any other attributes the generator added
public partial int Foo(...) {
    // Generator code
}
```

**Comment by yaakov-h** [comment, marked as answer]
doesn’t ConditionalAttribute already fill this need?

**Reply by mgravell** [reply]
Correction: maybe [Conditional ("DEBUG")] already would achieve this. I will investigate thanks!

(It comes down to: are they still in the semantic model when generators get them; I'll find out and revise accordingly)

**Reply by FaustVX** [reply]
`ConditionalAttribute` can only be applied on void-returning method and on class inheriting from `System.Attribute`.
So, I think that could works.

**Comment by mgravell** [comment]
Wow, @yaakov-h  is 100% correct - I can already do everything here just by adding `[Conditional("DEBUG")]` on my custom attributes - they still appear in the semantic model, so: life is good. As for inbuilt ones: that's entirely for the C# team, and I have no dog in that race. Awesome - nothing to do / see here, thanks!

For context / usage: see https://github.com/DapperLib/DapperAOT/commit/e3b3866fd0b02283634c10e6b97dbae31b349ca7
